src/main/java/frc/robot/subsystems/Swerve.java

```diff
@@ -0,0 +1,275 @@
+            .getStructTopic("Current pose", Pose2d.struct).publish();
+
+    private final SwerveDriveOdometry odometry;
+    private final KrakenSwerveModule[] modules = new KrakenSwerveModule[4];
+    private ChassisSpeeds chassisSpeeds = new ChassisSpeeds();
+
+    private boolean active = true;
+
+    public Swerve() {
+        ShuffleboardTab tab = Shuffleboard.getTab("Drivetrain");
+        for (int i = 0; i < modules.length; i++) {
+            modules[i] = new KrakenSwerveModule(
+                    tab.getLayout(DriveConstants.LAYOUT_TITLE[i], BuiltInLayouts.kList)
+                            .withSize(2, 4)
+                            .withPosition(i * 2, 0),
+                    DriveConstants.DRIVE_ID[i],
+                    DriveConstants.STEER_ID[i],
+                    DriveConstants.ENCODER_ID[i]);
+        }
+
+        // Load the RobotConfig from the GUI settings. You should probably
+        // store this in your Constants file
+        try{
+            config = RobotConfig.fromGUISettings();
+          } catch (Exception e) {
+            // Handle exception as needed
+            e.printStackTrace();
+          }
+
+        AutoBuilder.configure(
+                this::pose,
+                this::resetOdometry,
+                this::getChassisSpeeds,
+                (speeds, feedforwards) -> drive(speeds),
+                new PPHolonomicDriveController(
+                        new PIDConstants(AutoConstants.kPXController, 0.0, 0.0), // Translation PID constants
+                        new PIDConstants(AutoConstants.kPThetaController, 0, 0, 0.01) // Rotation PID constants
+                ),
+                config,
+                () -> {
+                    var alliance = DriverStation.getAlliance();
+                    if (alliance.isPresent()) {
+                        return alliance.get() == DriverStation.Alliance.Red;
+                    }
+                    return false;
+                },
+                this);
+
+        odometry = new SwerveDriveOdometry(kinematics, rotation(), modulePositions(),
+                new Pose2d(0, 0, new Rotation2d()));
+
+    }
+
+    public void zeroGyro() {
+        pigeon2.setYaw(0);
+    }
+
+    public Rotation2d rotation() {
+        return new Rotation2d(absoluteRotation());
+    }
+
+    public double absoluteRotation() {
+        double rotation = pigeon2.getYaw().getValueAsDouble() % 360;
+        rotation += (rotation < 0) ? 360 : 0;
+        return Math.toRadians(rotation);
+    }
+
+    public void drive(ChassisSpeeds chassisSpeeds) {
+        this.chassisSpeeds = chassisSpeeds;
+    }
+
+    public void stop() {
+        chassisSpeeds = new ChassisSpeeds();
+    }
+
+    public double distance(Pose2d reference_point) {
+        Transform2d dist = odometry.getPoseMeters().minus(reference_point);
+        return Math.sqrt((dist.getX() * dist.getX()) + (dist.getY() * dist.getY()));
```

Comment:
We can replace this with:
odometry.getPoseMeters().getTranslation().getDistance(reference_point.getTranslation);